### DISCUSSION TITLE
Sorting Supported Locales at Base Theme

### DISCUSSION THREAD

**Discussion by mibnd** [keycloak/keycloak#19191] [Ideas]
Hey!

Using FreeMarker's built-in `sort_by` function to sort the supported locales in `template.ftl` at the base theme is a great way to ensure that the list of locales is presented in a clear and organized manner. This will ensure that the locales are presented in a clear and organized manner, making it easier for users to select their preferred language.

What do you think about sorting supported locales in `template.ftl` by FreeMarker `sort_by` at base theme? Something like this:

```diff
<#-- theme/base/login/template.ftl -->
...
-<#list locale.supported as l>
+<#list locale.supported?sort_by("label") as l>
    <li class="${properties.kcLocaleListItemClass!}">
        <a class="${properties.kcLocaleItemClass!}" href="${l.url}">${l.label}</a>
    </li>
</#list>
...
```

```diff
<#-- theme/base/account/template.ftl -->
...
-<#list locale.supported as l>
+<#list locale.supported?sort_by("label") as l>
    <li class="kc-dropdown-item"><a href="${l.url}">${l.label}</a></li>
</#list>
...
```

**Comment by john-gom** [comment]
I work with another open source project that is adopting Keycloak and we have a very large number of locales, so sorting is essential. I have also found that the current picker is very difficult to use with a long list.

I'm thinking of creating a PR to change the locale selector to be a simple HTML select element as this is much clearer semantically (for accessibility) and works as expected on mobile devices, etc.

However, I'm aware that this kind of change could break existing UI customizations so would this sort of thing normally be done with a feature flag? Any input would be appreciated.

**Reply by jonkoops** [reply]
We started work on a new theme for the login based on [PatternFly 5](https://www.patternfly.org/), which includes a revised locale selector. If you want to try it out, you can build Keycloak from source, or download the [latest nightly](https://github.com/keycloak/keycloak/releases/tag/nightly), then pass `--features=login2` when starting the server and select the `keycloak.v2` theme in the Administration Console.

If you feel this can be enhanced, feel free to log an issue and open a PR.

**Reply by john-gom** [reply]
Thanks @jonkoops . I'm not finding an easy way to work on the login pages and ui-shared libraries without doing full builds. I found some instructions for the admin-ui but not sure how to do it for the account-ui. Are there any tips somewhere that I can refer to?

**Reply by john-gom** [reply]
dw, I think I have got it vaguely working, just used the same pattern for admin-ui for account-ui. Still having to manually build ui-shared if I make changes there though.

**Comment by john-gom** [comment]
Hi @jonkoops , @edewit . I had a quick look at this and from what I can make out you have taken the HTML that patternfly automatically generates and used this to create the template.ftl. Is that right?

I was hoping I could repeat what I did for #17776 and use a typeahead style if there are 10 or more options, but I'm not sure how well this will work. The simple alternative would be to use a native HTML select element rather than using patternfly. I have tried this out locally and it seems to work OK. I'll create a PR for the simple select option for discussion but happy to consider alternatives.

**Comment by john-gom** [comment]
Raised PR #27737

**Comment by john-gom** [comment]
Hi @jonkoops, @edewit. I have raised a couple of PRs (#27737 and #27726) but haven't heard anything back in over a week. Have I missed out anything?
